### PR TITLE
CB-1923 redbeams api improvements

### PR DESCRIPTION
**Pull request by gergopapi2**

### CODE REVIEW

File: redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/DatabaseServerV4Endpoint.java [hunk 1]
```diff
@@ -91,18 +98,32 @@ DatabaseServerV4Responses list(@NotNull @QueryParam("environmentId") String envi
...
+    @DELETE
+    @Path("/name/{name}")
     @Produces(MediaType.APPLICATION_JSON)
     @ApiOperation(value = DatabaseServerOpDescription.DELETE_BY_NAME, produces = MediaType.APPLICATION_JSON, notes = Notes.DATABASE_SERVER_NOTES,
-        nickname = "deleteDatabaseServer")
-    DatabaseServerV4Response delete(@NotNull @QueryParam("environmentId") String environmentId, @PathParam("name") String name);
+            nickname = "deleteDatabaseServerByName")
+    DatabaseServerV4Response deleteByName(@NotNull @QueryParam("environmentCrn") String environmentCrn, @PathParam("name") String name);
+
+//    @DELETE
```
Comment (bhavanki) [inline redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/DatabaseServerV4Endpoint.java]: Looks like this commented-out stuff can go

File: redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/base/DatabaseServerV4Identifiers.java [hunk 2]
```diff
@@ -1,45 +1,27 @@
...
-
-    public String getName() {
-        return name;
-    }
-
-    public void setName(String name) {
-        this.name = name;
-    }
+    @ApiModelProperty(value = CRN, required = true)
+    private String resourceCrn;
```
Comment (bhavanki) [inline redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/base/DatabaseServerV4Identifiers.java]: As a separate change, not now: This should be annotated with @ValidCrn. I bet there are other places the API should use it ...

File: redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/base/DatabaseServerV4Identifiers.java [hunk 3]
```diff
@@ -1,45 +1,27 @@
...
-import javax.validation.constraints.Size;
 
 import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
-import com.sequenceiq.redbeams.doc.ModelDescriptions;
-import com.sequenceiq.redbeams.doc.ModelDescriptions.DatabaseServer;
 
 import io.swagger.annotations.ApiModelProperty;
 
 @JsonIgnoreProperties(ignoreUnknown = true)
 public class DatabaseServerV4Identifiers implements Serializable {
```
Comment (bhavanki) [inline redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/base/DatabaseServerV4Identifiers.java]: Since this class now only holds a single field, should it be eliminated?

File: redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/provision/AbstractRedbeamsProvisionAction.java [hunk 4]
```diff
@@ -68,7 +69,7 @@ protected RedbeamsContext createFlowContext(FlowParameters flowParameters,
         CloudContext cloudContext = new CloudContext(dbStack.getId(), dbStack.getName(), dbStack.getCloudPlatform(), dbStack.getPlatformVariant(),
                 location, userName, userName, accountId);
         // FIXME must use CRN
```
Comment (bhavanki) [inline redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/provision/AbstractRedbeamsProvisionAction.java]: This comment can go away! 😃

File: redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/termination/AbstractRedbeamsTerminationAction.java [hunk 5]
```diff
@@ -76,7 +77,7 @@ protected RedbeamsContext createFlowContext(FlowParameters flowParameters,
             cloudContext = new CloudContext(dbStack.getId(), dbStack.getName(), dbStack.getCloudPlatform(), dbStack.getPlatformVariant(),
                     location, userName, accountId);
             // FIXME must use CRN
```
Comment (bhavanki) [inline redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/termination/AbstractRedbeamsTerminationAction.java]: This one too

File: redbeams/src/main/java/com/sequenceiq/redbeams/repository/DBStackRepository.java [hunk 6]
```diff
@@ -1,15 +1,19 @@
...
 import org.springframework.data.jpa.repository.JpaRepository;
 
+import com.sequenceiq.cloudbreak.auth.altus.Crn;
+import com.sequenceiq.redbeams.domain.stack.DBStack;
+
 @Transactional(Transactional.TxType.REQUIRED)
 public interface DBStackRepository extends JpaRepository<DBStack, Long> {
 
-    Optional<DBStack> findByNameAndEnvironmentId(String name, String environmentId);
+    Optional<DBStack> findByNameAndEnvironmentCrn(String name, String environmentId);
```
Comment (bhavanki) [inline redbeams/src/main/java/com/sequenceiq/redbeams/repository/DBStackRepository.java]: Does this work with the changed method name? I thought that JPA devised the query based on the method name. I could be wrong.

File: redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/database/request/CreateDatabaseV4Request.java [hunk 7]
```diff
@@ -18,11 +16,8 @@
 @JsonIgnoreProperties(ignoreUnknown = true)
 public class CreateDatabaseV4Request implements Serializable {
 
-    @ApiModelProperty(ENVIRONMENT_CRN)
-    private String environmentCrn;
-
     @ApiModelProperty(DatabaseServer.NAME)
```
[the remark this answers is not in the document]
Comment (gergopapi2) [reply redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/database/request/CreateDatabaseV4Request.java]: Thanks, done.

File: redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/base/DatabaseServerV4Identifiers.java (hunk 2 above)
Comment (gergopapi2) [reply redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/base/DatabaseServerV4Identifiers.java]: Jira for that: https://jira.cloudera.com/browse/CB-2452

File: redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/base/DatabaseServerV4Identifiers.java (hunk 3 above)
Comment (gergopapi2) [reply redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/base/DatabaseServerV4Identifiers.java]: Yes, thank you!

File: redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/termination/AbstractRedbeamsTerminationAction.java (hunk 5 above)
Comment (gergopapi2) [reply redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/termination/AbstractRedbeamsTerminationAction.java]: Done

File: redbeams/src/main/java/com/sequenceiq/redbeams/repository/DBStackRepository.java (hunk 6 above)
Comment (gergopapi2) [reply redbeams/src/main/java/com/sequenceiq/redbeams/repository/DBStackRepository.java]: In some of the entities (where there were no handcrafted queries) I did rename the field environmentId to environmentCrn, so this should work. 

On the other hand I am thinking we should retain environmentId, just to decouple the concept of an id from its representation as a crn. So I rename it back to environmentId. And, I think it would be better to rename this field for all or none of the entities but not just some.

File: redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/DatabaseServerV4Endpoint.java [hunk 8]
```diff
@@ -45,23 +43,22 @@
...
-    @Path("{nameOrCrn}")
-    @Produces(MediaType.APPLICATION_JSON)
-    @ApiOperation(value = DatabaseServerOpDescription.GET_BY_NAME_OR_CRN, produces = MediaType.APPLICATION_JSON, notes = Notes.DATABASE_SERVER_NOTES,
-            nickname = "getDatabaseServerByNameOrCrn")
-    DatabaseServerV4Response getByNameOrCrn(@NotNull @QueryParam("environmentId") String environmentId, @PathParam("nameOrCrn") String nameOrCrn);
+    DatabaseServerV4Responses list(@NotNull @QueryParam("environmentCrn") String environmentCrn);
 
     @GET
-    @Path("/crn/{crn}")
+    @Path("{crn}")
```
Comment (topolyai5) [inline redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/DatabaseServerV4Endpoint.java]: Use the same pattern as the other module, for example: `EnvironmentEndpoint`

File: redbeams/src/main/java/com/sequenceiq/redbeams/service/dbserverconfig/DatabaseServerConfigService.java [hunk 9]
```diff
@@ -76,7 +78,7 @@
     @Inject
     private UserGeneratorService userGeneratorService;
 
-    public Set<DatabaseServerConfig> findAll(Long workspaceId, String environmentId, Boolean attachGlobal) {
+    public Set<DatabaseServerConfig> findAll(Long workspaceId, String environmentId) {
```
Comment (topolyai5) [inline redbeams/src/main/java/com/sequenceiq/redbeams/service/dbserverconfig/DatabaseServerConfigService.java]: We don't use workspaceId in microservices as well

File: redbeams/src/main/java/com/sequenceiq/redbeams/domain/stack/DBStack.java [hunk 10]
```diff
@@ -33,6 +33,10 @@
     @SequenceGenerator(name = "dbstack_generator", sequenceName = "dbstack_id_seq", allocationSize = 1)
     private Long id;
 
+    @Convert(converter = CrnConverter.class)
+    @Column(nullable = false)
+    private Crn resourceCrn;
```
Comment (topolyai5) [inline redbeams/src/main/java/com/sequenceiq/redbeams/domain/stack/DBStack.java]: What is the reason for Crn object? We add more unnecessary conversation to the process.

File: redbeams/src/test/java/com/sequenceiq/redbeams/controller/v4/database/DatabaseV4ControllerTest.java [hunk 11]
```diff
@@ -133,8 +133,8 @@ public void testRegister() {
     // public void testTestWithName() {
```
Comment (topolyai5) [inline redbeams/src/test/java/com/sequenceiq/redbeams/controller/v4/database/DatabaseV4ControllerTest.java]: Could you please delete the commented lines if you don't need for these lines anymore?

File: redbeams/src/test/java/com/sequenceiq/redbeams/converter/spi/DBStackToDatabaseStackConverterTest.java [hunk 12]
```diff
@@ -22,6 +13,15 @@
 import org.junit.Test;
 import org.mockito.InjectMocks;
 
+import com.sequenceiq.cloudbreak.api.endpoint.v4.common.DatabaseVendor;
```
Comment (topolyai5) [inline redbeams/src/test/java/com/sequenceiq/redbeams/converter/spi/DBStackToDatabaseStackConverterTest.java]: Unnecessary changes

File: redbeams/src/test/java/com/sequenceiq/redbeams/domain/stack/DBStackTest.java [hunk 13]
```diff
@@ -2,16 +2,16 @@
...
-import com.sequenceiq.cloudbreak.auth.altus.Crn;
-import com.sequenceiq.cloudbreak.common.json.Json;
-import com.sequenceiq.redbeams.api.model.common.Status;
-
 import java.util.Map;
 
 import org.junit.Before;
 import org.junit.Test;
 
+import com.google.common.collect.ImmutableMap;
```
Comment (topolyai5) [inline redbeams/src/test/java/com/sequenceiq/redbeams/domain/stack/DBStackTest.java]: Unnecessary changes

File: redbeams/src/test/java/com/sequenceiq/redbeams/flow/redbeams/provision/AbstractRedbeamsProvisionActionTest.java [hunk 14]
```diff
@@ -4,6 +4,12 @@
 import static org.mockito.Mockito.when;
 import static org.mockito.MockitoAnnotations.initMocks;
 
+import org.junit.Before;
```
Comment (topolyai5) [inline redbeams/src/test/java/com/sequenceiq/redbeams/flow/redbeams/provision/AbstractRedbeamsProvisionActionTest.java]: Unnecessary changes

File: redbeams/src/test/java/com/sequenceiq/redbeams/flow/redbeams/termination/AbstractRedbeamsTerminationActionTest.java [hunk 15]
```diff
@@ -4,6 +4,14 @@
 import static org.mockito.Mockito.when;
 import static org.mockito.MockitoAnnotations.initMocks;
 
+import java.util.Optional;
```
Comment (topolyai5) [inline redbeams/src/test/java/com/sequenceiq/redbeams/flow/redbeams/termination/AbstractRedbeamsTerminationActionTest.java]: Unnecessary changes

File: redbeams/src/main/java/com/sequenceiq/redbeams/service/dbserverconfig/DatabaseServerConfigService.java (hunk 9 above)
Comment (gergopapi2) [reply redbeams/src/main/java/com/sequenceiq/redbeams/service/dbserverconfig/DatabaseServerConfigService.java]: Yes. It remained there because of historic reasons ... I do not want toe delete it as part of this jira but we could get rid of it. @bhavanki, what do you suggest?

File: redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/DatabaseServerV4Endpoint.java (hunk 8 above)
Comment (gergopapi2) [reply redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/DatabaseServerV4Endpoint.java]: Thanks.

First, EnvironmentEndpoint and e.g. SdxEndpoint use a somewhat different pattern. Namely, in sdx you have /{name} and /crn/{crn}.
Since in our case the main way to reach resources is via crn, we made the /{crn} the main way to get a resource and /name/{name the secondary. 

If you still think we should have /crn/{crn}, we can change.

File: redbeams/src/main/java/com/sequenceiq/redbeams/domain/stack/DBStack.java (hunk 10 above)
Comment (gergopapi2) [reply redbeams/src/main/java/com/sequenceiq/redbeams/domain/stack/DBStack.java]: Yes, string would be easier. 
This way, however, 
1) you can be sure that the stored string can be mapped to a crn, and
2) the type of the variable tells you what this thing is (crn), not the name.

File: redbeams/src/test/java/com/sequenceiq/redbeams/controller/v4/database/DatabaseV4ControllerTest.java (hunk 11 above)
Comment (gergopapi2) [reply redbeams/src/test/java/com/sequenceiq/redbeams/controller/v4/database/DatabaseV4ControllerTest.java]: I created a jira to work on the DatabaseConfig api part.
https://jira.cloudera.com/browse/CB-2480

File: redbeams/src/main/java/com/sequenceiq/redbeams/service/dbserverconfig/DatabaseServerConfigService.java (hunk 9 above)
Comment (bhavanki) [reply redbeams/src/main/java/com/sequenceiq/redbeams/service/dbserverconfig/DatabaseServerConfigService.java]: At this point, I'm fine getting rid of it. It's becoming more of a wart.

File: redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/DatabaseServerV4Endpoint.java (hunk 8 above)
Comment (bhavanki) [reply redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/DatabaseServerV4Endpoint.java]: We did discuss the path format for CRN vs name for redbeams. We looked at the environment and freeipa services, and found that they have different standards. When I look right now, environment uses /crn and /name, but freeipa doesn't use either - environments passed by CRN are passed to endpoints without "/crn". (If I recall correctly, when we looked earlier, freeipa was using /crn but not /name.)

So, we decided to not use "/crn" because the CRN fits the definition of a RESTful resource identifier, and it's architecturally better to focus on it as the primary identifier in the API. So, "/crn" would not be necessary.
